File: spikes/builtin-host-boundary/s2-shell/probe.mjs
// S2 probe (node) — what does `spawn(command, {shell:true})` actually run?
// That is what js/src/builtin.ts:163 does today, so this is the shipped
// behaviour, not a reimplementation. On Windows `shell:true` means
// `%COMSPEC% /d /s /c`, i.e. cmd.exe — a different language from `sh`.
import { spawn } from "node:child_process"

const probes = [
  ["posix_var", "echo $HOME"],
  ["bashism", "[[ -d . ]] && echo bashism-ok"],
  ["cmd_var", "echo %USERPROFILE%"],
  ["which_self", "echo $0"],
]

function run(cmd) {
  return new Promise((resolve) => {
    const child = spawn(cmd, { shell: true })
    let out = ""
    child.stdout.on("data", (d) => (out += d))
    child.stderr.on("data", (d) => (out += d))
    child.on("close", (code) => resolve({ code, out: out.trim() }))
    child.on("error", (e) => resolve({ code: -1, out: String(e) }))
  })
}

console.log(`platform=${process.platform} COMSPEC=${process.env.COMSPEC ?? "<unset>"}`)
for (const [name, cmd] of probes) {
  const { code, out } = await run(cmd)
  console.log(`${name}: exit=${code} out=${JSON.stringify(out)}`)
}

File: spikes/builtin-host-boundary/s2-shell/probe.py
# S2 probe (python) — what does `subprocess.run(cmd, shell=True)` run? That is
# what python/src/toolnexus/builtin.py:189 does today. On POSIX it is
# `/bin/sh -c`; on Windows it is `%COMSPEC% /c`, i.e. cmd.exe.
import os
import subprocess
import sys

PROBES = [
    ("posix_var", "echo $HOME"),
    ("bashism", "[[ -d . ]] && echo bashism-ok"),
    ("cmd_var", "echo %USERPROFILE%"),
    ("which_self", "echo $0"),
]

print(f"platform={sys.platform} COMSPEC={os.environ.get('COMSPEC', '<unset>')}")
for name, cmd in PROBES:
    try:
        p = subprocess.run(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, timeout=15)
        print(f"{name}: exit={p.returncode} out={p.stdout.decode(errors='replace').strip()!r}")
    except Exception as e:  # noqa: BLE001 - a probe reports, it does not handle
        print(f"{name}: raised {type(e).__name__}: {e}")
